[AIRFLOW-1552] Airflow Filter_by_owner not working with password_auth

thomasbrockmeier opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [x] Fix [AIRFLOW-1552](https://issues.apache.org/jira/browse/AIRFLOW-1552), Airflow filter_by_owner not working with password_auth

### Description

- [x]  Here are some details about my PR, including screenshots of any UI changes:

- `airflow.contrib.auth.backends.password_auth.load_user` would query `models.User` and cast the result to `PasswordUser` instead of querying `PasswordUser` directly. This would map the model's fields to a nested instance, e.g., `password_user_instance.user.superuser` instead of the expected `password_user_instance.superuser`.

- The `superuser` field is added to `models.User` in the form of a SQLAlchemy column, instead of a hard-coded boolean.

- The overridden `PasswordUser.is_superuser` method is removed, so the actual value in the database is referenced instead.

### Tests


### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- [x] Passes `flake8`

Commit:
[AIRFLOW-1552] - Query PasswordUser on password login

ashb commented on airflow/contrib/auth/backends/password_auth.py:
```diff
@@ -94,10 +94,6 @@ def data_profiling(self):
 
-    def is_superuser(self):
-        """Access all the things"""
-        return True
```
Why was this removed?

ashb replied:
Oh, just read the PR message, NM.

ashb commented on airflow/contrib/auth/backends/password_auth.py:
```diff
@@ -106,8 +102,8 @@ def load_user(userid, session=None):
 
-    user = session.query(models.User).filter(models.User.id == int(userid)).first()
-    return PasswordUser(user)
+    user = session.query(PasswordUser).filter(PasswordUser.id == int(userid)).first()
```
This change looks better, but I need to check it doesn't actually break other things - parts of the code may expect user.user to exist (as bonkers as that sounds)

ashb replied:
I suspect this will fail tests as https://github.com/apache/incubator-airflow/blob/b7a7fd66d693dbfbc471a6d08bc274441ee4841c/airflow/www/utils.py#L299 won't work anymore.

And this (admittedly silly) API is part of the external API that people have written custom auth backends against so we can't change it. If we were keeping these I'd say it's worth "fixing" this (as you have here) but since we're removing the custom auth backends in Favour of Flask-AppBuilder in 2.0.0 it's probably worth just keeping the sillyness.

thomasbrockmeier replied:
If it keeps the API from breaking, I can see if I can fix it in [views.py:2070](https://github.com/apache/incubator-airflow/blob/master/airflow/www/views.py#L2070) with something like

```
    do_filter = FILTER_BY_OWNER and (not current_user.user.is_superuser())
```

Not the nicest way to handle this, I guess, but perhaps worth considering if it enables filtering DAGs by owner for the time being?

Do you have any indication when 2.0.0 is scheduled for release? Depending on the time frame, I may be able to invest a couple of hours to look into this further as this feature would make my life a lot easier :)

ashb replied:
No timing on 2.0.0 (q1 or q2 next year) but the "RBAC" based UI is available behind a config flag on 1.10.0 so you can start experimenting with it already https://github.com/apache/incubator-airflow/blob/master/UPDATING.md#new-webserver-ui-with-role-based-access-control

thomasbrockmeier replied:
My use case unfortunately requires DAG level access control. I have tried running master, but the DAG level roles available there do not seem to work on my machine.

I have updated my code to leave the API alone, please let me know if there is anything else I can do to improve this PR.

Commit:
[AIRFLOW-1552] Airflow Filter_by_owner not working with password_auth

Querying on PasswordUser may break API. Query on User and cast to
PasswordUser like before, but now check against `password_user_instance.user.superuser`

ashb commented on airflow/migrations/versions/41f5f12752f8_add_superuser_field.py:
```diff
@@ -0,0 +1,43 @@
+
+
+def upgrade():
+    op.add_column('users', sa.Column('superuser', sa.Boolean(), default=False))
```
This probably needs a note in UPDATING.md as (I think?) without action this will make all current users not super-users?

ashb replied:
The default of False is the right thing to do though

thomasbrockmeier committed:
[AIRFLOW-1552] Airflow Filter_by_owner not working with password_auth

Add note to UPDATING.md

thomasbrockmeier committed:
[AIRFLOW-1552] Add migration description
